perf(query): reduce regex query fanout

alextheimer opened the pull request:
**Pull Request checklist**

- [x] The commit(s) message(s) follows the contribution [guidelines](CONTRIBUTING.md) ?
- [x] Tests for the changes have been added (for bug fixes / features) ?
- [x] Docs have been added / updated (for bug fixes / features) ?

Currently, the `ShardKeyRegexPlanner` asks its inner planner to create an `ExecPlan` for each one of a `LogicalPlan`'s shard keys. This is inefficient; if 100 shard keys are matched but all data lies on only two partitions, it would be much more efficient to instead ask the inner planner to materialize one plan per partition.

This PR materializes `ExecPlan`s in this more-efficient manner. The `ShardKeyRegexPlanner` now passes `LogicalPlans` with original regex filters to lower-level planners. The `MultiPartitionPlanner` now has infrastructure added to resolve shard-keys and determine the partitions to be queried; it now materializes plans per-partition (instead of per-shard-key). Again, regex is passed to lower-level planners; the `SingleClusterPlanner` now also has the infrastructure to materialize regex-filtered plans.

vishramachandran reviewed (commented):
@alextheimer  Query Stats need to be keyed with "multiple-ns" for the namespace field for NS-Regex queries within partition.

alextheimer commented:
@vishramachandran it looks like the latest `develop` has this covered (I've rebased the PR):
- In `TimeSeriesShard`: https://github.com/filodb/FiloDB/blob/6cb5433b202239fd1ac4baaf201468532444e8a2/core/src/main/scala/filodb.core/memstore/TimeSeriesShard.scala#L1936
- In `DownsampledTimeSeriesShard`: https://github.com/filodb/FiloDB/blob/6cb5433b202239fd1ac4baaf201468532444e8a2/core/src/main/scala/filodb.core/downsample/DownsampledTimeSeriesShard.scala#L355

`QueryStats` are also updated in e.g. label-value iterators, but I'll leave those keys alone since they aren't related to PromQL queries.

alextheimer commented on prometheus/src/main/scala/filodb/prometheus/ast/Vectors.scala:
```diff
@@ -241,9 +241,14 @@ sealed trait Vector extends Expression {
-                                ColumnFilter(labelMatch.label, query.Filter.EqualsRegex(labelValue))
+        case RegexMatch      =>
+          // Relax the length limit only for matchers that contain at most the "|" special character.
+          if (QueryUtils.containsUnescapedNonPipeRegexChars(labelValue)) {
```
Double-check equal with Lucene functionality.

alextheimer replied:
No longer applies.

alextheimer commented on coordinator/src/main/scala/filodb.coordinator/queryplanner/LogicalPlanUtils.scala:
```diff
@@ -422,7 +423,15 @@ object LogicalPlanUtils extends StrictLogging {
+      val resolvedFilters = shardKeyFilters.flatMap{ filters =>
+        val updFilters = filters.map { filter => filter.filter match {
+          case EqualsRegex(values: String) if !QueryUtils.containsUnescapedNonPipeRegexChars(values) =>
+            values.split('|').map(value => ColumnFilter(filter.column, Equals(value))).toSeq
```
Make work for escaped pipes

amolnayak311 commented on coordinator/src/test/scala/filodb.coordinator/queryplanner/MultiPartitionPlannerSpec.scala:
```diff
@@ -923,94 +923,82 @@ class MultiPartitionPlannerSpec extends AnyFunSpec with Matchers with PlanValida
-    stitchRvsExec.children(0).isInstanceOf[PromQlRemoteExec] shouldEqual (true)
-    stitchRvsExec.children(1).isInstanceOf[PromQlRemoteExec] shouldEqual (true)
-
+    execPlan.isInstanceOf[PromQlRemoteExec] shouldEqual (true)
```
Why should it generate just one? The previous one was generating two children,  both for different partition endpoints.

alextheimer replied:
Nope, they were generated for the same endpoints:
```
    val remoteExec = stitchRvsExec.children(0).asInstanceOf[PromQlRemoteExec]
    ...
    remoteExec.queryEndpoint shouldEqual "remote-url"

    val remoteExec2 = stitchRvsExec.children(1).asInstanceOf[PromQlRemoteExec]
    ...
    remoteExec2.queryEndpoint shouldEqual "remote-url"
```
This test used to generate two children because one child plan was materialized per shard-key.

alextheimer replied:
Some additional comments: this test needed an update because -- now that all split plans (plans with at least one shard-key that resides in >1 partitions) are materialized via `materializePlanHandleSplitLeaf`-- an empty result is returned when all timestamp evaluations require data from >1 partitions.

That would happen here:
- This is an instant query, so only one timestamp is evaluated at `t=1000s`.
- The lookback is `100s`.
- The partition split happens at `950s`.

I've pushed the start of the second partition back far enough to fit the entire `100s` lookback, and now we should only expect one `RemoteExec` to be returned.

amolnayak311 replied:
Got it, so the method would return ``EmptyResultExec`` if left unchanged? If yes why not keep that test and additionally test your case too which goes to just one partition for a timestamp >= 1050?

alextheimer replied:
Good point -- I will re-add the `EmptyResultExec` test.

amolnayak311 commented on coordinator/src/test/scala/filodb.coordinator/queryplanner/MultiPartitionPlannerSpec.scala:
```diff
@@ -923,94 +923,82 @@ class MultiPartitionPlannerSpec extends AnyFunSpec with Matchers with PlanValida
+    remoteExec.queryEndpoint shouldEqual "remote-url2"
   }
 
+//  // TODO: what is this the purpose of this test? The lookback is 300s, and the assignment
```
Why comment out the test?

alextheimer replied:
I was hoping others would know what this test was for-- it looks like we used to expect partition-split plans would always return at least one data-point per partition, regardless of lookback/offset. That doesn't jive with the way we handle partition-splits elsewhere in the planners, so I think this test should be removed.

The reason it passed prior to this PR is because the MPP sent `PeriodicSeries` and `RawSeries` through the old `materializePeriodicAndRawSeries` logic. That always materialized at least one plan per partition.

`materializePeriodicAndRawSeries` was built to materialize plans with equality-only filters, so that would not work for this PR. Instead, plans are:
- if split, passed to the regex-handling `materializeSplitLeafPlan`, which accounts correctly for lookback.
- if not split, forwarded as-is to any partitions that own the query data.

alextheimer commented on coordinator/src/main/scala/filodb.coordinator/queryplanner/MultiPartitionPlanner.scala:
```diff
@@ -96,7 +98,12 @@ class MultiPartitionPlanner(partitionLocationProvider: PartitionLocationProvider
       case lp: TsCardinalities                => materializeTsCardinalities(lp, qContext).plans.head
-      case _                                  => walkLogicalPlanTree(logicalPlan, qContext).plans.head
+      case _                                  =>
+        val result = walkLogicalPlanTree(logicalPlan, qContext)
```
Good point -- this was probably needed for some previous PR iteration, but tests pass without it, now. I will drop this change since the extra logic isn't necessary.

alextheimer commented on coordinator/src/main/scala/filodb.coordinator/queryplanner/MultiPartitionPlanner.scala:
```diff
@@ -389,64 +332,127 @@ class MultiPartitionPlanner(partitionLocationProvider: PartitionLocationProvider
+      case ps: PeriodicSeries => materializeNonSplitPeriodicAndRawSeries(ps, qContext)
+      case rcm: RawChunkMeta => materializeNonSplitPeriodicAndRawSeries(rcm, qContext)
+      case rs: RawSeries => materializeNonSplitPeriodicAndRawSeries(rs, qContext)
       case x => throw new IllegalArgumentException(s"unhandled type: ${x.getClass}")
```
I wish I knew this sooner, it bugged me that `match`es never threw compile-time errors:
> It is, however, possible to do this for sealed traits whose subclasses are all also at least sealed or final (thus effectively "turning off" the capability of loading new subclasses at runtime). In that case, Scala will optimize the pattern match to a more efficient representation, and you will get a warning for a non-exhaustive pattern match.

From [here](https://stackoverflow.com/a/65489322).

I will add a line for each subclass 👍

alextheimer commented on coordinator/src/main/scala/filodb.coordinator/queryplanner/MultiPartitionPlanner.scala:
```diff
@@ -389,64 +332,127 @@ class MultiPartitionPlanner(partitionLocationProvider: PartitionLocationProvider
-                                             .exists(getPartitions(_, qParams).size > 1)
+    // Create one plan per RawSeries/shard-key pair, then resolve its partitions.
+    // If any resides on more than one partition, the leaf is "split".
+    val hasMultiPartitionLeaves =
```
This is confusing because it should be renamed to `hasSplitPartitionLeaves`. "Multi-partition" implied "split" when shard-keys were filtered only by equality.

Detailed explanation for completeness: The old logic was
```
    val hasMultiPartitionLeaves = LogicalPlan.findLeafLogicalPlans(logicalPlan)
                                             .exists(getPartitions(_, qParams).size > 1)
```
Suppose some plan has regex shard-key filters. If the resolved keys reside in multiple partitions, the `getPartitions.size` step might return some value `> 1`. The old logic would interpret that as a partition split even though none might exist.

amolnayak311 commented on coordinator/src/test/scala/filodb.coordinator/queryplanner/ShardKeyRegexPlannerSpec.scala:
```diff
@@ -72,48 +78,33 @@ class ShardKeyRegexPlannerSpec extends AnyFunSpec with Matchers with ScalaFuture
-      contains(ColumnFilter("_ns_", Equals("App-2"))) shouldEqual(true)
-  }
-
-  it("should generate Exec plan for implicit ws query") {
```
@alextheimer if this test doesn't break your work leave it there. There are some changes in ``ShardKeyRegexPlanner`` that this test asserts

alextheimer replied:
[discussed offline]

alextheimer commented on coordinator/src/main/scala/filodb.coordinator/queryplanner/QueryUtils.scala:
```diff
@@ -0,0 +1,35 @@
+/**
+ * Storage for miscellaneous utility functions.
+ */
+object QueryUtils {
```
Much cleaner -- I've added almost exactly this (but with SeqExtractors: https://stackoverflow.com/a/19147469)
